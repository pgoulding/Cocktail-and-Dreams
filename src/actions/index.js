export const addIngredients = (ingredients) => ({
  type:'ADD_INGREDIENTS',
  ingredients
})

export const removeIngredient= (ingredient) => ({
  type:'REMOVE_INGREDIENT',
  ingredient
})

export const addFavorite = (cocktail) => ({
  type:'ADD_FAVORITE',
  cocktail
})

export const removeFavorite = (drinkId) => ({
  type:'REMOVE_FAVORITE',
  drinkId
})


