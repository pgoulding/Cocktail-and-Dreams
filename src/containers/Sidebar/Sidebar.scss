aside {
  background-color: rgb(70, 70, 70);
  padding:10px;
  width:240px;
  float:left;
  height: 100vh;
  color: #fff;
  h3{
    text-align: center;
  }
  form{
    margin: 5px;
    p{
      text-align: center;
      margin: 5px;
    }
    button{
      width:100%;
      background-color: #2356FF;
      line-height: 1.5em;
      font-size: 2vw;
      text-shadow: 0 0 3vw rgb(189, 189, 189);
      color:rgb(199, 199, 199);
      font-weight: 700;
      &:hover{
        color:#fff;
      }
    }
    select{
      width:100%;
      height:220px;
      font-size: 2vh;
      border:none;
      option{
        border:1px solid rgb(199, 199, 199);
        &:hover{
          background-color: rgb(199, 199, 199);
          color:#000;
        }
      }
    }
  }
}

.ingredient__select-output {
  height: 150px;
  overflow: scroll;
  border:2px solid rgb(180, 180, 180);
  margin:5px;
  button{
    border:1px solid rgb(58, 0, 83);
    outline: 0;
    margin: 2px;
    color: black;
    font-size: 1rem;
    &:hover{
      background-color: rgb(110, 110, 110);
      color:white;
    }
    // text-shadow: 1px 1px 1px black;
    // background-color: violet;
    border-radius: 4px;
  }
}