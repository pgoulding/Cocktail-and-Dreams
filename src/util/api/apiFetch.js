import {drinksApiKey} from '../apiKey'

const listURL = `https://www.thecocktaildb.com/api/json/v1/${drinksApiKey}/list.php?`
const searchURL = `https://www.thecocktaildb.com/api/json/v1/${drinksApiKey}/search.php?`

export const getIngredientList = async () => {
  try {
    const response = await fetch(`${listURL}i=list`)
    const parsed = await response.json()
    return parsed.drinks
  } catch (error) {
    throw Error(error.message)
  }
}

export const getIngredientDetails = async (ingredientName) => {
  try {
    const response = await fetch(`${searchURL}i=${ingredientName}`)
    const parsed = await response.json()
    return parsed
  } catch (error) {
    throw Error(error.message)
  }
}

export const findDrinkbyName = async (drinkName) => {
  try {
    const response = await fetch(`${searchURL}s=${drinkName}`)
    if(response.ok) {
      const parsed = await response.json()
      return parsed
    } else {
      throw Error('Failed to find drink.')
    }
  } catch (error) {
    throw Error (error.message)
}
}



