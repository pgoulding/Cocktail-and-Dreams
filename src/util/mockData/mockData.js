export const ingredientsList = [
  {"strIngredient1": "Dark rum"},
  {"strIngredient1": "Sweet Vermouth"},
  {"strIngredient1": "Strawberry schnapps"},
  {"strIngredient1": "Scotch"},
  {"strIngredient1": "Apricot brandy"},
  {"strIngredient1": "Triple sec"},
  {"strIngredient1": "Apple brandy"},
  {"strIngredient1": "Carbonated water"},
  {"strIngredient1": "Chocolate syrup"},
]

export const mockUserIngredients =[
  {
    "idIngredient": "571",
    "strIngredient": "151 proof rum",
    "strDescription": "Bacardi 151 is a brand of highly alcoholic rum made by Bacardi Limited of Hamilton, Bermuda.",
    "strType": "Rum"
  },
{
    "idIngredient": "573",
    "strIngredient": "Absinthe",
    "strDescription": "Absinthe is historically described as a distilled, highly alcoholic (45–74% ABV / 90–148 U.S. proof) beverage.[1][2][3][4] It is an anise-flavoured spirit derived from botanicals, including the flowers and leaves of Artemisia absinthium (\"grand wormwood\"), together with green anise, sweet fennel, and other medicinal and culinary herbs.",
    "strType": "Spirit"
  },
  {
    "idIngredient": "8",
    "strIngredient": "Absolut Vodka",
    "strDescription": "Absolut Vodka is a brand of vodka, produced near Åhus, in southern Sweden.",
    "strType": "Vodka"
  },
{
    "idIngredient": "18",
    "strIngredient": "Amaretto",
    "strDescription": "Amaretto (Italian for \"a little bitter\") is a sweet, almond-flavoured, Italian liqueur associated with Saronno, Italy. Various commercial brands are made from a base of apricot pits, almonds, or both. When served as a beverage.",
    "strType": "Liqueur"
  },
{
    "idIngredient": "221",
    "strIngredient": "Fruit Punch",
    "strDescription": null,
    "strType": null
  },
{
    "idIngredient": "457",
    "strIngredient": "Sour Mix",
    "strDescription": "Sour mix (also known as sweet and sour mix) is a mixer that is yellow-green in color and is used in many cocktails",
    "strType": "Soft Drink"
  }
]

export const mockDrink = {
  "idDrink": "11007",
  "strDrink": "Margarita",
  "strAlcoholic": "Alcoholic",
  "strGlass": "Cocktail glass",
  "strInstructions": "Rub the rim of the glass with the lime slice to make the salt stick to it. Take care to moisten only the outer rim and sprinkle the salt on it. The salt should present to the lips of the imbiber and never mix into the cocktail. Shake the other ingredients with ice, then carefully pour into the glass.",
  "strDrinkThumb": "https://www.thecocktaildb.com/images/media/drink/wpxpvu1439905379.jpg",
  "strIngredient1": "Tequila",
  "strIngredient2": "Triple sec",
  "strIngredient3": "Lime juice",
  "strMeasure1": "1 1/2 oz ",
  "strMeasure2": "1/2 oz ",
  "strMeasure3": "1 oz ",
  "strCreativeCommonsConfirmed": "No",
  "dateModified": "2015-08-18 14:42:59"
}

export const mockCleanedDrink = {
  drinkId: "11007",
  drinkName: "Margarita",
  alcoholic: "Alcoholic",
  glass: "Cocktail glass",
  instructions: "Rub the rim of the glass with the lime slice to make the salt stick to it. Take care to moisten only the outer rim and sprinkle the salt on it. The salt should present to the lips of the imbiber and never mix into the cocktail. Shake the other ingredients with ice, then carefully pour into the glass.",
  favorited: false,
  drinkImageSource: "https://www.thecocktaildb.com/images/media/drink/wpxpvu1439905379.jpg",
  ingredients:[
    {ingredient: "Tequila", amount:"4oz"},
    {ingredient:"Mix", amount:"8oz"},
  ]
}