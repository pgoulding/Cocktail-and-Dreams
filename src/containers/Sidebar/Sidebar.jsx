import React, { Component } from 'react'
import { getIngredientList, getIngredientDetails } from '../../util/api/apiFetch'
import { connect } from 'react-redux'
import { addIngredients } from '../../actions/index'
import './Sidebar.scss'
export class Sidebar extends Component {
  constructor() {
    super()
    this.state={
      ingredientList:[],
      selectedIngredients:[]
    }
  }

  async componentDidMount(){
    const ingredientList = await getIngredientList()
    const alphabetizedList = ingredientList.map(item => item.strIngredient1).sort()
    this.setState({ingredientList: alphabetizedList})
  }

  sendIngredientsToStore =async (e) => {
    e.preventDefault()
    const ingredientsList = await this.state.selectedIngredients
      .map(async ingredient => getIngredientDetails(ingredient))
    const returned = await Promise.all(ingredientsList)
    this.props.addIngredients(returned)
    this.setState({selectedIngredients:[]})
  }

  handleChange= (e) => {
    if(!this.state.selectedIngredients.includes(e.target.value)) {
      this.setState({ selectedIngredients: [...this.state.selectedIngredients, e.target.value] });
    } else {
      this.setState({selectedIngredients:[...this.state.selectedIngredients.filter(ingredient => ingredient !== e.target.value)]})
    }
  }

  removeItem=(e) => {
    e.preventDefault()
    this.setState({selectedIngredients:[...this.state.selectedIngredients.filter(item => item !==e.target.value)]})
  }

  optionsBlock = () => {
    return this.state.ingredientList.map(ingredient => {
      return <option value={ingredient}>{ingredient}</option>
    })
  }

  selectedOption = () => {
    return this.state.selectedIngredients.map(selected => {
      return <button value={selected} id={selected}className="single-ingredient-button" onClick={(e)=> this.removeItem(e)}>{selected}</button>
    })
  }

  render() { 
    return (
      <aside>
        <h3>Select Ingredient</h3>
        <div className='ingredient__select-output'>
            {this.selectedOption()}
        </div>
        <form>
          <p> Ingredients list (scroll)</p>
          <select className="ingredients-list-select" value={this.state.selectedIngredients} multiple onChange={(e) => this.handleChange(e)}>
            {this.optionsBlock()}
          </select>
          <button className="send-ingredients-button" onClick={(e)=> this.sendIngredientsToStore(e)}>Add Ingredients</button>
        </form>
      </aside>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  addIngredients: (ingredients) => dispatch(addIngredients(ingredients))
})

export default connect(null, mapDispatchToProps)(Sidebar)
