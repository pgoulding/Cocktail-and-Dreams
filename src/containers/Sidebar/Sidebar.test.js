import { Sidebar} from './Sidebar'
import { mockDrink, mockUserIngredients, mockCleanedDrink } from '../../util/mockData/mockData';
import { shallow } from 'enzyme'
import React from 'react';
import {mapDispatchToProps} from './Sidebar'
import {addIngredients} from '../../actions/index'
import {getIngredientList} from '../../util/api/apiFetch'

jest.mock('../../util/api/apiFetch', () => ({
  getIngredientList: jest.fn().mockImplementation(() => {
    return [{ "strIngredient1": "Dark rum" },
    { "strIngredient1": "Sweet Vermouth" },
    { "strIngredient1": "Strawberry schnapps" },
    { "strIngredient1": "Scotch" },
    { "strIngredient1": "Apricot brandy" },
    { "strIngredient1": "Triple sec" },
    { "strIngredient1": "Apple brandy" },
    { "strIngredient1": "Carbonated water" },
    { "strIngredient1": "Chocolate syrup" }];
  })
}));

describe('Sidebar', () => {

  let wrapper;
  let e;
  beforeEach(() =>{
    wrapper = shallow(<Sidebar addIngredients={jest.fn()}/>)
    e = {
      preventDefault: jest.fn(),
      target: {
        name: 'cocktailName',
        value: 'margarita'
      }
    }
  })
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it.skip('should call getIngredientList on CDM', () => {
    let wrapper = shallow(<Sidebar addIngredients={jest.fn()} />)
    wrapper.instance().getIngredientList = jest.fn()
    wrapper.instance().componentDidMount()
    expect(wrapper.instance().getIngredientList).toHaveBeenCalled()
    // I don't know how to get this to work
  })

  it('should map dispatch to props', () => {
    const mockDispatch = jest.fn()
    const actionToDispatch = addIngredients(mockUserIngredients)
    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.addIngredients(mockUserIngredients)
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  })

  it('should handleChange', ()=> {
    wrapper.instance().handleChange = jest.fn();
    const mockEvent = { preventDefault: jest.fn() }
    wrapper.find('.ingredients-list-select').simulate('change', mockEvent);
    expect(wrapper.instance().handleChange).toHaveBeenCalled();
  })

  it('should call sendIngredientsToStore on button click', () => {
    wrapper.instance().sendIngredientsToStore = jest.fn()
    const mockEvent = { preventDefault: jest.fn() }
    wrapper.find('.send-ingredients-button').simulate('click', mockEvent)
    expect(wrapper.instance().sendIngredientsToStore).toHaveBeenCalled()
  })

  it.skip('should remove an item when clicked on', () =>{
    wrapper.instance().componentDidMount()
    wrapper.find('#Scotch').simulate('click', e)
    expect(wrapper.state('ingredientList'))
    //how can we test for length here?
  })

  it('should add a item when clicked on', () => {
    
  })

})