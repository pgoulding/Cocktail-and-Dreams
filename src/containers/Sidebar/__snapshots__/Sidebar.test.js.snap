// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should match the snapshot 1`] = `ShallowWrapper {}`;
